pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,210 @@
+	CreationTimestamp time.Time
+	// The local clock when the Pod is actually deleted from kube-apiserver. Nil if it's not deleted yet.
+	DeletionTimestamp *time.Time
+}
+
+// NewPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func NewPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPods:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		prevPods:     cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+		clock:        clock,
+		timestampMap: map[string]*podLocalTimestamps{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			DeleteFunc: s.onPodDelete,
+		})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+		return
+	}
+	err := s.curPods.Add(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when adding pod to curPods")
+		return
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	s.timestampMap[string(pod.UID)] = &podLocalTimestamps{CreationTimestamp: timeNow}
+	klog.V(4).InfoS("Process Pod Create Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing Delete Event")
+			return
+		}
+	}
+	err := s.prevPods.Add(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when adding pod to prevPods")
+		return
+	}
+	err = s.curPods.Delete(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when deleting pod from curPods")
+		return
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	s.timestampMap[string(pod.UID)].DeletionTimestamp = &timeNow
+	s.podsToDelete.AddAfter(pod, DelayTime)
+	klog.V(4).InfoS("Process Pod Delete Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool) {
+	// Query current Pods first
+	curPod, ok := s.getPods(s.curPods, ip, startTime)
+	if ok {
+		return curPod, true
+	}
+	// Query previous Pods
+	return s.getPods(s.prevPods, ip, startTime)
```

Comment:
It's probably better *not* to release the mutex in between the 2 indexer lookups. I feel like `getPods` should also be a closure inside `GetPodByIPAndTime`, rather than a standalone `PodStore` method.

Reply:
Fixed. Thanks.